configuration.md

```diff
@@ -0,0 +1,433 @@
+
+Multiple Graphite listeners may be configured on unique ports and associated
+with different check UUIDs. See the section on [Graphite
+ingestion](graphite-ingestion.md) for details.
+
+##### Graphite listener address
+
+The IP address on which to listen, or the special `*` to listen on any local IP
+address.
+
+Default: `*`
+
+##### Graphite listener port
+
+The TCP port number to listen on.
+
+Default: 2003
+
+##### Graphite listener type
+
+The type of listener. IRONdb implements a Graphite-compatible handler in
+libmtev, using the custom type "graphite".
+
+Default: graphite
+
+##### Graphite listener config
+
+These configuration items control which check UUID, name, and account ID are
+associated with this listener. The first Graphite listener is configured during
+[initial installation](installation.md).
+* `check_uuid` is the identifier for all metrics ingested via this listener.
+* `check_name` is a meaningful name that is used in
+  [namespacing](graphite-ingestion.md#namespacing).
+* `account_id` is also part of namespacing, for disambiguation.
+
+#### CLI listener
+
+```
+<listener address="127.0.0.1" port="32322" type="mtev_console">
+  <config>
+    <line_protocol>telnet</line_protocol>
+  </config>
+</listener>
+```
+
+The CLI listener provides a local [telnet
+console](http://circonus-labs.github.io/libmtev/operations/telnet_console.html)
+for interacting with libmtev subsystems, including modifying configuration. As
+there is no authentication mechanism available for this listener, it is
+recommended that it only be operated on the localhost interface.
+
+##### CLI listener address
+
+The IP address on which to listen, or the special `*` to listen on any local IP
+address.
+
+Default: 127.0.0.1
+
+##### CLI listener port
+
+The TCP port number to listen on.
+
+Default: 32322
+
+##### CLI listener type
+
+The CLI listener uses the built-in libmtev type "mtev_console" to allow access
+to the telnet console.
+
+Default: mtev_console
+
+### pools
+
+```
+<pools>
+  <nnt_put concurrency="16"/>
+  <raw_writer concurrency="4"/>
+</pools>
+```
+
+Resource pools within IRONdb are used for various functions, such as reading
+and writing metric data. Some aspects of pool behavior are configurable,
+typically to adjust the amount of concurrency within a given pool.
+
+The defaults presented are widely applicable to most workloads, but may be
+adjusted to improve throughput. Use caution when raising these values too high,
+as it could produce thrashing and _decrease_ performance.
+
+#### pools nnt_put concurrency
+
+The number of job queues used for writing to numeric rollup files. Writes to a
+given rollup file will always occur in the same queue.
+
+Default: the number of physical CPU cores present during installation
+
+#### pools raw_writer concurrency
+
+The number of job queues used for writing to the raw metrics database.
+Additionally, IRONdb will use 4x this number of job queues for reading from the
+raw metrics database.
+
```

Comment:
I think we need to add more pools as default here, or enumerate what pools you can use.

AppNexus is using:
```
  <pools>
    <rollup concurrency="8" />
    <nnt_put concurrency="8"/>
    <raw_writer concurrency="4"/>
    <raw_surrogate_writer concurrency="4" />
    <rest_graphite_numeric_get concurrency="100"/>
    <rest_graphite_find_metrics concurrency="100"/>
    <rest_graphite_fetch_metrics concurrency="100"/>
  </pools>
```

Reply:
I didn't know about these others; I was basing all of this on our default config.  I would like to get these in there too.  For all the ones I didn't have before, are these the in-code defaults?

Reply:
They are not in code defaults.  Defaults are:

```
<raw_surrogate_writer concurrency="1" />
<rollup concurrency="1" />
<rest_graphite_numeric_get concurrency="4"/> <!-- master jobq on graphite fetches -->
<rest_graphite_find_metrics concurrency="4"/> <!-- for finding names -->
<rest_graphite_fetch_metrics concurrency="10"/> <!-- for doing local and remote actual fetches -->
```

Reply:
OK so the above config looks to _improve_ performance by default -- I like it!

Reply:
So should we just be explicit about the in-code defaults or should we ship with higher concurrency "out of the box"?

Reply:
The defaults serve most people well.. I won't want to encourage people to make 100 concurrency queues as they will attempt to do this on a single CPU GCP VM and it will seize up from thread contention and context switches.

Reply:
Sounds good, will document the code defaults.